pkg/operator/ceph/object/user/controller.go

```diff
@@ -124,6 +129,78 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 		return err
 	}
 
+	// watch secrets referenced by CephObjectStoreUser.spec.keys
+	const (
+		secretNameField = "spec.keys.secretNames"
+	)
+
+	err = mgr.GetFieldIndexer().IndexField(context.TODO(), &cephv1.CephObjectStoreUser{TypeMeta: controllerTypeMeta}, secretNameField, func(obj client.Object) []string {
+		var secretNames []string
+		for _, k := range obj.(*cephv1.CephObjectStoreUser).Spec.Keys {
+			if k.AccessKeyRef != nil && k.AccessKeyRef.Name != "" {
+				secretNames = append(secretNames, k.AccessKeyRef.Name)
+			}
+			if k.SecretKeyRef != nil && k.SecretKeyRef.Name != "" {
+				secretNames = append(secretNames, k.SecretKeyRef.Name)
+			}
+		}
+		slices.Sort(secretNames)
+		return slices.Compact(secretNames)
+	})
+	if err != nil {
+		logger.Errorf("failed to setup IndexField for CephObjectStoreUser.Spec.Keys. %v", err)
+		return err
+	}
+
+	// Always trigger a reconcile when a secret is deleted. This will cause a
+	// reconciliation failure to happen immediately in hopes of alerting the end
+	// user to the configuration problem.
+	changedOrDeleted := predicate.Or(
+		predicate.TypedResourceVersionChangedPredicate[*corev1.Secret]{},
+		predicate.TypedFuncs[*corev1.Secret]{
+			DeleteFunc: func(e event.TypedDeleteEvent[*corev1.Secret]) bool {
+				return true
+			},
+		},
+	)
+
+	err = c.Watch(
+		source.Kind(
+			mgr.GetCache(),
+			&corev1.Secret{},
+			handler.TypedEnqueueRequestsFromMapFunc(
```

Comment:
Coming from [this discussion](https://github.com/rook/rook/pull/15498/files#r2008117785), I think it's preferable to use the non-Typed method here to keep consistency with the rest of Rook's code.

Overall, I don't care if we use Typed or non-Typed, but I think/hope that keeping consitency across the codebase will be the most impactful for our lives as maintainers. If we want to update all controllers to use Typed methods in a future PR, that would be fine with me

Reply:
Consistency is a compelling argument.  This morning, I already started looking a converting the entire operator over to the controller-runtime generic types.

Would you be willing to take this as it currently is if I agree to immediate convert the rest of the code base over?  That would let me spend time on refactoring the other controllers rather than redoing this logic.

Reply:
Seems fine by me

Reply:
I made lots of progress on #15563 today.  I believe this PR is ready for review/merge as-is.  Is it blocked on #15563 being merged first?

Reply:
We can continue with this first

Reply:
https://github.com/rook/rook/pull/15563 is ready for review.